addons/account/models/account_move.py

```diff
@@ -3552,6 +3555,8 @@ def field_will_change(line, field_name):
                     raise UserError(_("You cannot edit the following fields due to restrict mode being activated on the journal: %s.") % ', '.join(INTEGRITY_HASH_LINE_FIELDS))
                 if any(key in vals for key in ('tax_ids', 'tax_line_ids')):
                     raise UserError(_('You cannot modify the taxes related to a posted journal item, you should reset the journal entry to draft to do so.'))
+                if 'product_id' in vals:
+                    raise ValidationError(_("You cannot modify a journal entry linked to a posted payment."))
```

Comment:
Can you try with something like this?
```python
    @property
    def _posted_protected_fields(self):
        return ['product_id', 'account_id', 'debit', 'credit', 'tax_ids', 'tax_line_ids']

    def write(vals):
        for line in self:
            if line.parent_state != 'draft' and vals.get('parent_state') != 'draft':
                protected_override = set(vals) & set(line._posted_protected_fields)
                if protected_override:
                    raise ValidationError(_(
                        "These fields cannot be changed on posted entries: %s",
                        get_nice_string(protected_override)
                    ))
```

The CI might not be happy about it, but this constraint definitely makes sense to me.

_________________

If it doesn't work, the message you copy pasted is not right. It is talking about a payment, but it is not related to payments at all.